neurow/lib/neurow/observability/system_stats.ex

```diff
@@ -0,0 +1,115 @@
+defmodule Neurow.Observability.SystemStats do
+  use Prometheus.Metric
+
+  def setup() do
+    Gauge.declare(
+      name: :memory_usage,
+      help: "Memory usage"
+    )
+
+    Boolean.declare(
+      name: :stopping,
+      help: "The node is currently stopping"
+    )
+
+    Gauge.set([name: :memory_usage], 0)
+    Boolean.set([name: :stopping], false)
+
+    Periodic.start_link(
+      run: fn -> Gauge.set([name: :memory_usage], :recon_alloc.memory(:usage)) end,
+      every: :timer.seconds(10)
+    )
+  end
+
+  def report_shutdown() do
+    Boolean.set([name: :stopping], true)
+  end
+
+  defmodule ProcessesStats do
+    defstruct [
+      :name_or_initial_func,
+      :current_func,
+      process_count: 0,
+      memory: 0,
+      message_queue_len: 0
+    ]
+  end
+
+  # Group process by name or initial function, and current function
+  # Then sort by memory usage and return the top consuming processes groups
+  def process_groups(result_count \\ 50) do
+    Process.list()
+    |> Enum.reduce(%{}, fn pid, acc ->
+      {name_or_initial_func, current_func} = grouping_attributes(pid)
+
+      process_info =
+        Process.info(pid, [
+          :memory,
+          :message_queue_len
+        ])
+
+      current_stats =
+        acc[{name_or_initial_func, current_func}] ||
+          %ProcessesStats{
+            name_or_initial_func: name_or_initial_func,
+            current_func: current_func
+          }
+
+      process_count = current_stats.process_count + 1
+      memory = current_stats.memory + (process_info[:memory] || 0)
+
+      message_queue_len =
+        current_stats.message_queue_len + (process_info[:message_queue_len] || 0)
+
+      Map.put(acc, {name_or_initial_func, current_func}, %ProcessesStats{
+        name_or_initial_func: name_or_initial_func,
+        process_count: process_count,
+        current_func: current_func,
+        memory: memory,
+        message_queue_len: message_queue_len
+      })
```

Comment:
**Suggestion** 🤔 : why not use [Map.update](https://hexdocs.pm/elixir/1.12/Map.html#update/4) here ? It takes a function that already have the previous state or default values